sharktuner/boo_tuner.py

```diff
@@ -0,0 +1,124 @@
+        parser = mio.get_miopen_parser()
+        parser.add_argument("-t")
+        parser.add_argument("--iter")
+        sig = mio.get_signature(parser.parse_args(cli_args))
+        if args.tmp_dir:
+            tmp_dir = Path(args.tmp_dir)
+            # Make sure directory is empty.
+            shutil.rmtree(tmp_dir, ignore_errors=True)
+            os.mkdir(tmp_dir)
+        else:
+            tmp_dir = Path(tempfile.mkdtemp(dir=".", prefix="boo-tuner-"))
+        boo_cache_dir = tmp_dir / "boo_cache"
+
+        # Run BOO compilation an extract source IR.
+        with boo_runtime.use_cache_dir(boo_cache_dir):
+            get_launchable(sig)(*sig.get_sample_args(device="cuda", seed=123))
+        [op_cache_dir] = os.listdir(boo_cache_dir)
+        [compile_command_path] = [
+            f
+            for f in os.listdir(boo_cache_dir / op_cache_dir)
+            if f.startswith("compile_command")
+        ]
+        with open(boo_cache_dir / op_cache_dir / compile_command_path) as f:
+            compile_command = f.read().strip()
+
+        # Re-compile to dump dispatches.
+        bench_dir = tmp_dir / "bench"
+        compile_args = shlex.split(compile_command) + [
+            "--iree-codegen-tuning-spec-path=",  # disable tuning spec flag as it prevents 'iree-config-add-tuner-attributes' from working.
+            "--iree-config-add-tuner-attributes",
+            "--iree-hal-dump-executable-benchmarks-to",
+            str(bench_dir),
+            "-o",
+            "/dev/null",
+        ]
+        print(f"> {shlex.join(compile_args)}")
+        subprocess.run(compile_args)
+
+        # Retain temporary directory if it was specified.
+        should_cleanup_tmp_dir = not args.tmp_dir
+        for bench_file in os.listdir(bench_dir):
+            bench_path = bench_dir / bench_file
+
+            # Run tuner.
+            tuner_args = [
+                "model-unused",
+                str(bench_path),
+                *(
+                    ("--starter-td-spec", str(starter_td_spec))
+                    if args.starter_td_spec is not None
+                    else ()
+                ),
+                *("--output-td-spec", str(args.output_td_spec)),
+                f"--devices={args.devices}",
+                "--model-tuner-num-dispatch-candidates=100",
+                f"--num-candidates={args.num_candidates}",
+                "--codegen-pipeline=llvmgpu_tile_and_fuse",
+                "--stop-after=benchmark-dispatches",
+            ]
+            print(f"> {shlex.join(tuner_args)}")
+            try:
+                tuner_main(tuner_args)
+                starter_td_spec = args.output_td_spec
+            except Exception as err:
+                # Print error and continue.
+                traceback.print_exception(err)
+                # Retain temporary directory for debugging.
+                should_cleanup_tmp_dir = False
+        if should_cleanup_tmp_dir:
+            shutil.rmtree(tmp_dir)
```

Comment:
Can we use a with statement to take care of the cleanup instead of relying on the control flow reaching this?